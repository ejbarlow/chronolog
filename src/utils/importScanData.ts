import path from "path";
import ScanProps from "../types/ScanProps";
import id from "./generateId";

async function importScanDataFromManifest(manifestPath: string) {
  const response = await fetch(manifestPath);
  const data = await response.json();
  const scanRoot = path.dirname(manifestPath);
  const scans: ScanProps[] = [];
  for (const vol in data) {
    for (const date in data[vol]) {
      for (const scan in data[vol][date]) {
        const fileStr: string = data[vol][date][scan];
        const dirStr = `${scanRoot}/${vol}/${date}`;
        const [y, m, d] = date.split("_").map((num: string) => parseInt(num));
        const dateProp = new Date(y, m - 1, d);
        const [, pagesStr] = fileStr.split("_");
        const pages = pagesStr.split("-").map((num: string) => parseInt(num));
        scans.push({
          uid: id(),
          path: `${dirStr}/${fileStr}`,
          thumbPath: `${dirStr}/${fileStr.replace(/\./, "_thumb.")}`,
          volume: parseInt(vol),
          pages,
          date: dateProp,
          version: 0,
        });
      }
    }
  }
  const highestPage: number = scans.reduce((pNum, { pages }) => {
    return pages[pages.length - 1] > pNum ? pages[pages.length - 1] : pNum;
  }, 0);
  return { scans, highestPage };
}

export { importScanDataFromManifest as default };
