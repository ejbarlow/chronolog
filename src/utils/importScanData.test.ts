import importScanData from "./importScanData";

const mockManifest = {
  "1": {
    "2022_01_01": [
      "JRNL_00-01_0001.webp",
      "JRNL_02-03_0001.webp",
      "JRNL_04-05_0001.webp",
      "JRNL_06-07_0001.webp",
      "JRNL_08-09_0001.webp",
      "JRNL_10-11_0001.webp",
    ],
    "2022_01_02": [
      "JRNL_04-05_0001.webp",
      "JRNL_06-07_0001.webp",
      "JRNL_10-11_0001.webp",
      "JRNL_12-13_0001.webp",
    ],
  },
  "2": {
    "2022_01_02": [
      "JRNL_00-01_0001.webp",
      "JRNL_02-03_0001.webp",
      "JRNL_04-05_0001.webp",
      "JRNL_06-07_0001.webp",
      "JRNL_08-09_0001.webp",
    ],
    "2022_01_03": [
      "JRNL_02-03_0001.webp",
      "JRNL_10-11_0001.webp",
      "JRNL_12-13_0001.webp",
    ],
  },
};

test("Creates valid scan data from manifest", () => {
  jest.spyOn(global, "fetch").mockImplementation((): Promise<any> => {
    return Promise.resolve({
      json() {
        return Promise.resolve(mockManifest);
      },
    });
  });
  importScanData("public/manifest.json").then((scanData) => {
    expect(scanData).toMatchSnapshot();
  });
});
