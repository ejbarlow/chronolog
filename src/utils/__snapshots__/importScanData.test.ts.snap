// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creates valid scan data from manifest 1`] = `
Array [
  Object {
    "date": 2022-01-01T00:00:00.000Z,
    "pages": Array [
      0,
      1,
    ],
    "path": "public/1/2022_01_01/JRNL_00-01_0001.webp",
    "thumbPath": "public/1/2022_01_01/JRNL_00-01_0001_thumb.webp",
    "uid": "CL_000000",
    "version": 0,
    "volume": 1,
  },
  Object {
    "date": 2022-01-01T00:00:00.000Z,
    "pages": Array [
      2,
      3,
    ],
    "path": "public/1/2022_01_01/JRNL_02-03_0001.webp",
    "thumbPath": "public/1/2022_01_01/JRNL_02-03_0001_thumb.webp",
    "uid": "CL_000001",
    "version": 0,
    "volume": 1,
  },
  Object {
    "date": 2022-01-01T00:00:00.000Z,
    "pages": Array [
      4,
      5,
    ],
    "path": "public/1/2022_01_01/JRNL_04-05_0001.webp",
    "thumbPath": "public/1/2022_01_01/JRNL_04-05_0001_thumb.webp",
    "uid": "CL_000002",
    "version": 0,
    "volume": 1,
  },
  Object {
    "date": 2022-01-01T00:00:00.000Z,
    "pages": Array [
      6,
      7,
    ],
    "path": "public/1/2022_01_01/JRNL_06-07_0001.webp",
    "thumbPath": "public/1/2022_01_01/JRNL_06-07_0001_thumb.webp",
    "uid": "CL_000003",
    "version": 0,
    "volume": 1,
  },
  Object {
    "date": 2022-01-01T00:00:00.000Z,
    "pages": Array [
      8,
      9,
    ],
    "path": "public/1/2022_01_01/JRNL_08-09_0001.webp",
    "thumbPath": "public/1/2022_01_01/JRNL_08-09_0001_thumb.webp",
    "uid": "CL_000004",
    "version": 0,
    "volume": 1,
  },
  Object {
    "date": 2022-01-01T00:00:00.000Z,
    "pages": Array [
      10,
      11,
    ],
    "path": "public/1/2022_01_01/JRNL_10-11_0001.webp",
    "thumbPath": "public/1/2022_01_01/JRNL_10-11_0001_thumb.webp",
    "uid": "CL_000005",
    "version": 0,
    "volume": 1,
  },
  Object {
    "date": 2022-01-02T00:00:00.000Z,
    "pages": Array [
      4,
      5,
    ],
    "path": "public/1/2022_01_02/JRNL_04-05_0001.webp",
    "thumbPath": "public/1/2022_01_02/JRNL_04-05_0001_thumb.webp",
    "uid": "CL_000006",
    "version": 0,
    "volume": 1,
  },
  Object {
    "date": 2022-01-02T00:00:00.000Z,
    "pages": Array [
      6,
      7,
    ],
    "path": "public/1/2022_01_02/JRNL_06-07_0001.webp",
    "thumbPath": "public/1/2022_01_02/JRNL_06-07_0001_thumb.webp",
    "uid": "CL_000007",
    "version": 0,
    "volume": 1,
  },
  Object {
    "date": 2022-01-02T00:00:00.000Z,
    "pages": Array [
      10,
      11,
    ],
    "path": "public/1/2022_01_02/JRNL_10-11_0001.webp",
    "thumbPath": "public/1/2022_01_02/JRNL_10-11_0001_thumb.webp",
    "uid": "CL_000008",
    "version": 0,
    "volume": 1,
  },
  Object {
    "date": 2022-01-02T00:00:00.000Z,
    "pages": Array [
      12,
      13,
    ],
    "path": "public/1/2022_01_02/JRNL_12-13_0001.webp",
    "thumbPath": "public/1/2022_01_02/JRNL_12-13_0001_thumb.webp",
    "uid": "CL_000009",
    "version": 0,
    "volume": 1,
  },
  Object {
    "date": 2022-01-02T00:00:00.000Z,
    "pages": Array [
      0,
      1,
    ],
    "path": "public/2/2022_01_02/JRNL_00-01_0001.webp",
    "thumbPath": "public/2/2022_01_02/JRNL_00-01_0001_thumb.webp",
    "uid": "CL_000010",
    "version": 0,
    "volume": 2,
  },
  Object {
    "date": 2022-01-02T00:00:00.000Z,
    "pages": Array [
      2,
      3,
    ],
    "path": "public/2/2022_01_02/JRNL_02-03_0001.webp",
    "thumbPath": "public/2/2022_01_02/JRNL_02-03_0001_thumb.webp",
    "uid": "CL_000011",
    "version": 0,
    "volume": 2,
  },
  Object {
    "date": 2022-01-02T00:00:00.000Z,
    "pages": Array [
      4,
      5,
    ],
    "path": "public/2/2022_01_02/JRNL_04-05_0001.webp",
    "thumbPath": "public/2/2022_01_02/JRNL_04-05_0001_thumb.webp",
    "uid": "CL_000012",
    "version": 0,
    "volume": 2,
  },
  Object {
    "date": 2022-01-02T00:00:00.000Z,
    "pages": Array [
      6,
      7,
    ],
    "path": "public/2/2022_01_02/JRNL_06-07_0001.webp",
    "thumbPath": "public/2/2022_01_02/JRNL_06-07_0001_thumb.webp",
    "uid": "CL_000013",
    "version": 0,
    "volume": 2,
  },
  Object {
    "date": 2022-01-02T00:00:00.000Z,
    "pages": Array [
      8,
      9,
    ],
    "path": "public/2/2022_01_02/JRNL_08-09_0001.webp",
    "thumbPath": "public/2/2022_01_02/JRNL_08-09_0001_thumb.webp",
    "uid": "CL_000014",
    "version": 0,
    "volume": 2,
  },
  Object {
    "date": 2022-01-03T00:00:00.000Z,
    "pages": Array [
      2,
      3,
    ],
    "path": "public/2/2022_01_03/JRNL_02-03_0001.webp",
    "thumbPath": "public/2/2022_01_03/JRNL_02-03_0001_thumb.webp",
    "uid": "CL_000015",
    "version": 0,
    "volume": 2,
  },
  Object {
    "date": 2022-01-03T00:00:00.000Z,
    "pages": Array [
      10,
      11,
    ],
    "path": "public/2/2022_01_03/JRNL_10-11_0001.webp",
    "thumbPath": "public/2/2022_01_03/JRNL_10-11_0001_thumb.webp",
    "uid": "CL_000016",
    "version": 0,
    "volume": 2,
  },
  Object {
    "date": 2022-01-03T00:00:00.000Z,
    "pages": Array [
      12,
      13,
    ],
    "path": "public/2/2022_01_03/JRNL_12-13_0001.webp",
    "thumbPath": "public/2/2022_01_03/JRNL_12-13_0001_thumb.webp",
    "uid": "CL_000017",
    "version": 0,
    "volume": 2,
  },
]
`;
